torch/autograd/functional.py

```diff
@@ -0,0 +1,719 @@
+import torch
+
+# Utility functions
+
+def _as_tuple(inp, arg_name, fn_name):
+    # Ensures that the inputs is a tuple of Tensors
+    # Returns whether or not the original inp was a tuple and the tupled version of the input
+    tuple_inp = True
+    if not isinstance(inp, tuple):
+        inp = (inp,)
+        tuple_inp = False
+
+    for i, el in enumerate(inp):
+        if not torch.is_tensor(el):
+            raise TypeError("The {} given to {} must be either a Tensor or a tuple of Tensors but the"
+                            " value at index {} has type {}.".format(arg_name, fn_name, i, type(el)))
+
+    return tuple_inp, inp
+
+def _tuple_postprocess(inp, tuple_inp):
+    # Unpacks a potentially nested tuple of Tensors
+    # It is used to invert _as_tuple before returning to the user
+    if isinstance(tuple_inp, tuple):
+        assert len(tuple_inp) == 2
+        if not tuple_inp[1]:
+            inp = tuple(el[0] for el in inp)
+        if not tuple_inp[0]:
+            inp = inp[0]
+    else:
+        if not tuple_inp:
+            inp = inp[0]
+    return inp
+
+def _grad_preprocess(inputs, create_graph, need_graph):
+    # Preprocess the inputs to make sure they require gradient
+    # inputs is a tuple of Tensor to preprocess
+    # create_graph specifies if the user wants gradients to flow back to the Tensors in inputs
+    # need_graph specifies if we internally want gradients to flow back to the Tensors in res
+    # Note that we *always* create a new Tensor object to be able to see the difference between
+    # inputs given as arguments and the same Tensors automatically captured by the user function.
+    # Check this issue for more details on how that can happen: https://github.com/pytorch/pytorch/issues/32576
+    res = []
+    for inp in inputs:
+        if create_graph and inp.requires_grad:
+            # Use .view_as() to get a new Tensor in a differentiable way
+            res.append(inp.view_as(inp))
+        else:
+            res.append(inp.detach().requires_grad_(need_graph))
+    return tuple(res)
+
+
+def _grad_postprocess(inputs, create_graph):
+    # Postprocess the generated Tensors to avoid returning Tensors with history when the user does not
```

Comment:
nit: "when the user does not requested it" -> "when the user did not request it"